closed/make/Docs.gmk

```diff
@@ -20,4 +20,93 @@
+# Define a javadoc Group for the OpenJ9 specific modules
+# Module names are added to DOCS_MODULES in closed/make/common/Modules.gmk
+OPENJ9_GROUP_NAME := OpenJ9
+OPENJ9_GROUP_MODULES := openj9.*
+OPENJ9_GROUP_DESCRIPTION := \
+    Modules whose names start with {@code openj9.} contain APIs provided by the \
+    <a href="$(OPENJ9_JAVADOC_BASE_URL)">Eclipse OpenJ9</a> project. \
+    These APIs can only be used with a JDK which includes the OpenJ9 virtual machine. \
+    #
+JDK_GROUPS += OPENJ9
+
+################################################################################
+# Add openj9 specific modules to the list of modules for which javadoc is required
+OPENJ9_MODULES := $(filter openj9.%,$(call FindAllModules))
+DOCS_MODULES += $(OPENJ9_MODULES)
+
+OPENJ9_JAVADOC_MARKER := $(SUPPORT_OUTPUTDIR)/docs/_javadoc_JDK_API_openj9.marker
+JDK_API_CUSTOM_TARGETS += $(OPENJ9_JAVADOC_MARKER)
+
+OPENJ9_JAVADOC_SCRIPT := $(SUPPORT_OUTPUTDIR)/docs/_javadoc_openj9_footer_replace.sh
+OPENJ9_JAVADOC_TEMP  := $(SUPPORT_OUTPUTDIR)/docs/_javadoc_temp.html
+
+################################################################################
+# The OpenJ9 classes are built from the j9jcl_sources and closed/src
+# directories.
+# The packages are beneath directories called /share/classes (or /unix/classes,
+# /windows/classes etc.) which do not appear in the javadoc tree structure.
+# The commands below replace the footer in the javadoc html pages for the
+# following files:
+# 1. The openj9 java source files which override the openjdk versions
+# 2. The openj9.xxxx modules, including the javadoc generated files such as
+#    enum class pages, module-summary.html, package-summary.html etc.)
+# 3. The com/ibm/xxxx packages, including the javadoc generated files such as
+#    enum class pages, module-summary.html, package-summary.html etc.)
+# A temporary script file is used to avoid xargs command line length issues
+# on some platforms. 
+$(OPENJ9_JAVADOC_MARKER): $(DOCS_OUTPUTDIR)/api/index.html
+	$(ECHO) 'if [ -f $(DOCS_OUTPUTDIR)/api/$$1 ]; then $(CP) $(DOCS_OUTPUTDIR)/api/$$1 $(OPENJ9_JAVADOC_TEMP); $(OPENJ9_SUBST_COMMAND) <$(OPENJ9_JAVADOC_TEMP) >$(DOCS_OUTPUTDIR)/api/$$1; $(RM) $(OPENJ9_JAVADOC_TEMP); fi' >$(OPENJ9_JAVADOC_SCRIPT)
+	$(CD) $(SUPPORT_OUTPUTDIR)/j9jcl_sources; \
+		find . -type f -name "*.java" \
+			| sed 's|/[a-z][a-z]*/classes||g' \
+			| sed 's|\.java$$|\.html|g' \
+			| xargs -I {} sh $(OPENJ9_JAVADOC_SCRIPT) {}
+	$(CD) $(TOPDIR)/closed/src; \
+		find . -type f -name "*.java" \
+			| sed 's|/[a-z][a-z]*/classes||g' \
+			| sed 's|\.java$$|\.html|g' \
+			| xargs -I {} sh $(OPENJ9_JAVADOC_SCRIPT) {}
+	$(CD) $(DOCS_OUTPUTDIR)/api; \
+		find . -type f -path "*/openj9*/*.html" \
+			| xargs -I {} sh $(OPENJ9_JAVADOC_SCRIPT) {}
+	$(CD) $(DOCS_OUTPUTDIR)/api; \
+		find . -type f -path "*/com/ibm*/*.html" \
+			| xargs -I {} sh $(OPENJ9_JAVADOC_SCRIPT) {}
+	$(TOUCH) $@
```

Comment:
* This should use macros BASH, FIND, SED and XARGS instead of sh, find, sed and xargs, respectively.
* The helper script can process more than one file at a time so xargs doesn't need to spawn as many shells; capturing it in a make variable makes it easier to read.
* SED can apply more than one substitution per invocation. The `g` suffix should not be used - each of the substitutions should occur at most once per line of input. We should be matching `/ibm/`, not `/ibm*/`.
* FIND can search for more than one pattern per invocation.
* The helper script can be removed once we're done using it.

The result is this:
```
OPENJ9_JAVADOC_SCRIPT_BODY = \
    for path in "$$@" ; do \
      if [ -f "$(DOCS_OUTPUTDIR)/api/$$path" ] ; then \
        $(CP) "$(DOCS_OUTPUTDIR)/api/$$path" "$(OPENJ9_JAVADOC_TEMP)" ; \
        $(OPENJ9_SUBST_COMMAND) < "$(OPENJ9_JAVADOC_TEMP)" > "$(DOCS_OUTPUTDIR)/api/$$path" ; \
        $(RM) "$(OPENJ9_JAVADOC_TEMP)" ; \
      fi ; \
    done \
    #

$(OPENJ9_JAVADOC_MARKER) : $(DOCS_OUTPUTDIR)/api/index.html
	$(ECHO) '$(OPENJ9_JAVADOC_SCRIPT_BODY)' >$(OPENJ9_JAVADOC_SCRIPT)
	$(CD) $(SUPPORT_OUTPUTDIR)/j9jcl_sources ; \
		$(FIND) . -type f -name '*.java' \
			| $(SED) -e 's|/[^/]\+/classes/|/|' -e 's|\.java$$|.html|' \
			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
	$(CD) $(TOPDIR)/closed/src ; \
		$(FIND) . -type f -name '*.java' \
			| $(SED) -e 's|/[^/]\+/classes/|/|' -e 's|\.java$$|.html|' \
			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
	$(CD) $(DOCS_OUTPUTDIR)/api ; \
		$(FIND) . -type f '(' -path '*/openj9*/*.html' ')' -o '(' -path '*/com/ibm/*.html' ')' \
			| $(XARGS) $(BASH) $(OPENJ9_JAVADOC_SCRIPT)
	$(RM) -f $(OPENJ9_JAVADOC_SCRIPT)
	$(TOUCH) $@
```

Reply:
Changed